### PR TITLE
[gql] fix last queries for transaction connection

### PR DESCRIPTION
**Pull request by wlmyng**
## Description 

Currently, a graphql query for last 3 before 4 would return 3, 2, 1. This does not follow cursor specs. This PR addresses the bug to return 1, 2, 3, and adds a test to validate some scenarios.

## Test Plan 

transaction_connection/pagination.move

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-graphql-rpc/src/context_data/pg_backend.rs [hunk 1]
```diff
@@ -91,139 +91,134 @@ impl GenericQueryBuilder<Pg> for PgQueryBuilder {
...
+            ));
+        }
+
+        if let Some(input_object) = filter.input_object {
+            let subquery = tx_input_objects::dsl::tx_input_objects
+                .filter(tx_input_objects::dsl::object_id.eq(input_object.into_vec()))
+                .select(tx_input_objects::dsl::tx_sequence_number);
+
+            query = query.filter(transactions::dsl::tx_sequence_number.eq_any(subquery));
+        }
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/pg_backend.rs:214]: nit: new line

File: crates/sui-graphql-rpc/src/context_data/pg_backend.rs [hunk 2]
```diff
@@ -91,139 +91,134 @@ impl GenericQueryBuilder<Pg> for PgQueryBuilder {
...
+        if let Some(recipient) = filter.recv_address {
+            let subquery = tx_recipients::dsl::tx_recipients
+                .filter(tx_recipients::dsl::recipient.eq(recipient.into_vec()))
+                .select(tx_recipients::dsl::tx_sequence_number);
+
+            query = query.filter(transactions::dsl::tx_sequence_number.eq_any(subquery));
+        }
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/pg_backend.rs:201]: nit: whitespace

File: crates/sui-graphql-rpc/src/context_data/pg_backend.rs [hunk 3]
```diff
@@ -91,139 +91,134 @@ impl GenericQueryBuilder<Pg> for PgQueryBuilder {
...
             }
-        };
+        } else if let Some(sender) = filter.sent_address {
+            let subquery = tx_senders::dsl::tx_senders
+                .filter(tx_senders::dsl::sender.eq(sender.into_vec()))
+                .select(tx_senders::dsl::tx_sequence_number);
+
+            query = query.filter(transactions::dsl::tx_sequence_number.eq_any(subquery));
+        }
+        if let Some(recipient) = filter.recv_address {
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/pg_backend.rs:195]: nit: newline

File: crates/sui-graphql-rpc/src/context_data/pg_backend.rs [hunk 4]
```diff
@@ -91,139 +91,134 @@ impl GenericQueryBuilder<Pg> for PgQueryBuilder {
...
+            (Some(p), Some(m), Some(f)) => {
+                let subquery = tx_calls::dsl::tx_calls
+                    .filter(tx_calls::dsl::package.eq(p.into_vec()))
+                    .filter(tx_calls::dsl::module.eq(m))
+                    .filter(tx_calls::dsl::func.eq(f))
+                    .select(tx_calls::dsl::tx_sequence_number);
+
+                query = query.filter(transactions::dsl::tx_sequence_number.eq_any(subquery));
             }
+            _ => {}
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/context_data/pg_backend.rs:167]: Poor line, it looks lonely.

**Review by stefan-mysten — APPROVED** [review]

File: crates/sui-graphql-e2e-tests/tests/transaction_connection/pagination.move [hunk 5]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//# init --simulator --accounts A
+
+//# programmable --sender A --inputs 1 @A
```
Comment (amnn) [inline crates/sui-graphql-e2e-tests/tests/transaction_connection/pagination.move:6]: nit: can you differentiate the transactions some other way than their cursor? (e.g. use a different split amount in each one and display that in the transaction?) Cursors are going to be obfuscated soon.

File: crates/sui-graphql-rpc/src/context_data/pg_backend.rs (hunk 4 above)
Comment (amnn) [reply crates/sui-graphql-rpc/src/context_data/pg_backend.rs:167]: Could make it explicit that it's the `(None, None, None)` case, and leave a comment that it's supposed to be a `nop`?

```suggestion
            (None, None, None) => { /* nop */ }
```

File: crates/sui-graphql-rpc/src/context_data/pg_backend.rs [hunk 6]
```diff
@@ -91,139 +91,134 @@ impl GenericQueryBuilder<Pg> for PgQueryBuilder {
...
+        }
+        if let Some(recipient) = filter.recv_address {
+            let subquery = tx_recipients::dsl::tx_recipients
+                .filter(tx_recipients::dsl::recipient.eq(recipient.into_vec()))
+                .select(tx_recipients::dsl::tx_sequence_number);
+
+            query = query.filter(transactions::dsl::tx_sequence_number.eq_any(subquery));
+        }
+        if filter.paid_address.is_some() {
+            return Err(Error::Internal(
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/context_data/pg_backend.rs:203]: nit: Can you add a TODO here?

**Review by amnn — APPROVED** [review]
Nice!

File: crates/sui-graphql-e2e-tests/tests/transaction_connection/pagination.move (hunk 5 above)
Comment (wlmyng) [reply crates/sui-graphql-e2e-tests/tests/transaction_connection/pagination.move:6]: yes good idea

**Comment by wlmyng** [comment]
@amnn since we have your cursor framework, I'll close this PR so we can use that instead

**Comment by amnn** [comment]
Sounds good @wlmyng, I am working on a generic pagination algorithm as well, which should handle this case and proper previous/next page detection as well.

**Pull request closed without merging**